### PR TITLE
Add support for Recovery Code multi-factor authentication

### PR DESCRIPTION
**Pull request by lbalmaceda**
### Changes

This endpoint is used to authenticate when MFA is required and the user has a saved recovery code. A new recovery code might be sent back by the server, which needs to be displayed to the end-user.

### References

- See https://auth0.com/docs/api/authentication#verify-with-recovery-code
- Also normalized the line endings following: https://www.aleksandrhovhannisyan.com/blog/crlf-vs-lf-normalizing-line-endings-in-git/

### Testing

Please describe how this can be tested by reviewers. Be specific about anything not tested and reasons why. Since this library has unit testing, tests should be added for new functionality and existing tests should complete without errors.

- [x] This change adds unit test coverage

- [ ] This change adds integration test coverage

- [ ] This change has been tested on the latest version of the platform/language or why not

### Checklist

- [x] I have read the [Auth0 general contribution guidelines](https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md)

- [x] I have read the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md)

- [x] All existing and new tests complete without errors

### CODE REVIEW

File: auth0/src/main/java/com/auth0/android/result/Credentials.kt [hunk 1]
```diff
@@ -62,4 +62,13 @@ public open class Credentials(
...
     internal open val currentTimeInMillis: Long
         get() = System.currentTimeMillis()
+
+    /**
+     * Getter for the new multi-factor authentication recovery code. Only available if these credentials are the result of logging in using an MFA recovery code.
+     *
+     * @return the new MFA recovery code.
+     */
+    @field:SerializedName("recovery_code")
+    public var recoveryCode: String? = null
```
Comment (lbalmaceda) [inline auth0/src/main/java/com/auth0/android/result/Credentials.kt:72]: having a setter separate from the constructor prevents introducing a breaking change, as verified with the tests

File: auth0/src/main/java/com/auth0/android/request/internal/CredentialsDeserializer.kt [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        val idToken = context.deserialize<String>(jsonObject.remove("id_token"), String::class.java)
+        val accessToken =
+            context.deserialize<String>(jsonObject.remove("access_token"), String::class.java)
+        val type = context.deserialize<String>(jsonObject.remove("token_type"), String::class.java)
+        val refreshToken =
+            context.deserialize<String>(jsonObject.remove("refresh_token"), String::class.java)
+        val expiresIn = context.deserialize<Long>(jsonObject.remove("expires_in"), Long::class.java)
+        val scope = context.deserialize<String>(jsonObject.remove("scope"), String::class.java)
+        val recoveryCode =
+            context.deserialize<String>(jsonObject.remove("recovery_code"), String::class.java)
```
Comment (lbalmaceda) [inline auth0/src/main/java/com/auth0/android/request/internal/CredentialsDeserializer.kt:32]: lines added here, reading the `recovery_code` property. The rest is just java->kotlin conversion from the existing class

File: auth0/src/test/java/com/auth0/android/request/internal/CredentialsDeserializerMock.kt [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+        accessToken: String,
+        type: String,
+        refreshToken: String?,
+        expiresAt: Date,
+        scope: String?,
+        recoveryCode: String?
+    ): Credentials {
+        val credentials =
+            CredentialsMock(idToken, accessToken, type, refreshToken, expiresAt, scope)
+        credentials.recoveryCode = recoveryCode
```
Comment (lbalmaceda) [inline auth0/src/test/java/com/auth0/android/request/internal/CredentialsDeserializerMock.kt:19]: java->kotlin converstion to be able to call this internal setter

File: auth0/src/test/java/com/auth0/android/result/CredentialsTest.kt [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+            Credentials("idToken", "accessToken", "type", "refreshToken", Date(), "openid profile")
+        MatcherAssert.assertThat(credentials.scope, Matchers.`is`("openid profile"))
+    }
+
+    @Test
+    public fun shouldGetRecoveryCode() {
+        val date = Date()
+        val credentials: Credentials =
+            CredentialsMock("idToken", "accessToken", "type", "refreshToken", date, "scope")
+        credentials.recoveryCode = "recoveryCode"
```
Comment (lbalmaceda) [inline auth0/src/test/java/com/auth0/android/result/CredentialsTest.kt:35]: java->kotlin converstion to be able to call this internal setter

File: auth0/src/main/java/com/auth0/android/result/Credentials.kt [hunk 5]
```diff
@@ -62,4 +62,13 @@ public open class Credentials(
...
+        internal set
 }
```
Comment (Widcket) [inline auth0/src/main/java/com/auth0/android/result/Credentials.kt:74]:
```suggestion
}

```

File: auth0/src/main/java/com/auth0/android/request/internal/CredentialsDeserializer.kt [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        scope: String?,
+        recoveryCode: String?
+    ): Credentials {
+        val credentials = Credentials(
+            idToken, accessToken, type, refreshToken, expiresAt, scope
+        )
+        credentials.recoveryCode = recoveryCode
+        return credentials
+    }
+}
```
Comment (Widcket) [inline auth0/src/main/java/com/auth0/android/request/internal/CredentialsDeserializer.kt:68]:
```suggestion
}

```

File: auth0/src/main/java/com/auth0/android/result/Credentials.kt (hunk 5 above)
Comment (Widcket) [inline auth0/src/main/java/com/auth0/android/result/Credentials.kt:74]:
```suggestion
}

```

File: auth0/src/test/java/com/auth0/android/result/CredentialsTest.kt [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        MatcherAssert.assertThat(credentials.recoveryCode, Matchers.`is`("recoveryCode"))
+    }
+}
```
Comment (Widcket) [inline auth0/src/test/java/com/auth0/android/result/CredentialsTest.kt:38]:
```suggestion
}

```

File: auth0/src/main/java/com/auth0/android/request/internal/CredentialsDeserializer.kt (hunk 6 above)
Comment (lbalmaceda) [reply auth0/src/main/java/com/auth0/android/request/internal/CredentialsDeserializer.kt:68]: These files have been applied the default code formatting shortcut to match the same style we follow in the rest of the files.

Comment (Widcket) [reply auth0/src/main/java/com/auth0/android/request/internal/CredentialsDeserializer.kt:68]: It's more about following the UNIX convention of ending files with a newline.

Comment (Widcket) [reply auth0/src/main/java/com/auth0/android/request/internal/CredentialsDeserializer.kt:68]: (text files)

File: auth0/src/test/java/com/auth0/android/authentication/AuthenticationAPIClientTest.kt [hunk 8]
```diff
@@ -172,6 +172,39 @@ public class AuthenticationAPIClientTest {
...
+        val client = AuthenticationAPIClient(auth0)
+        client.loginWithRecoveryCode("ey30.the-mfa-token.value", "123456")
+            .start(callback)
+        ShadowLooper.idleMainLooper()
+        assertThat(
+            callback, AuthenticationCallbackMatcher.hasPayloadOfType(
+                Credentials::class.java
+            )
+        )
+        assertThat(callback.payload.recoveryCode, Matchers.`is`("654321"))
```
Comment (Widcket) [inline auth0/src/test/java/com/auth0/android/authentication/AuthenticationAPIClientTest.kt:189]: If we're asserting the recovery code on the payload, should we be asserting the token value as well?

Comment (Widcket) [reply auth0/src/test/java/com/auth0/android/authentication/AuthenticationAPIClientTest.kt:189]: Unless that is asserted in some other test, but the same could be said about the `Accept-Language` assert.

Comment (lbalmaceda) [reply auth0/src/test/java/com/auth0/android/authentication/AuthenticationAPIClientTest.kt:189]: @Widcket This line checks for the content of the `response`. Here we only care about checking that if a new recovery code was sent, we can parse it. (see one is 123456 and the other is 654321).

A few lines below we assert the content of the `request`, which includes both mfa token and recovery code.

File: auth0/src/test/java/com/auth0/android/result/CredentialsTest.kt [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+        val credentials =
+            Credentials("idToken", "accessToken", "type", "refreshToken", Date(), "openid profile")
+        MatcherAssert.assertThat(credentials.scope, Matchers.`is`("openid profile"))
+    }
+
+    @Test
+    public fun shouldGetRecoveryCode() {
+        val date = Date()
+        val credentials: Credentials =
+            CredentialsMock("idToken", "accessToken", "type", "refreshToken", date, "scope")
```
Comment (Widcket) [inline auth0/src/test/java/com/auth0/android/result/CredentialsTest.kt:34]: Shouldn't `recoveryCode` be passed to the constructor?

**Review by Widcket — APPROVED** [review]
LGTM, only have a question left that is not a blocker.

File: auth0/src/test/java/com/auth0/android/result/CredentialsTest.kt (hunk 9 above)
Comment (lbalmaceda) [reply auth0/src/test/java/com/auth0/android/result/CredentialsTest.kt:34]: There is no new constructor that takes a recovery code. I didn't want to create a secondary constructor just for this one property, and changing the existing one would be a breaking change even when default values are used. Kotlin doesn't enforce the use of "named arguments" so anyone could be calling it the way we are calling it in this specific line and would be facing a breaking change due to a missing property.

**Pull request merged**